nifi-extension-bundles/nifi-github-bundle/nifi-github-extensions/src/main/java/org/apache/nifi/github/GitHubRepositoryClient.java

```diff
@@ -0,0 +1,404 @@
+                        .collect(Collectors.toSet());
+            } catch (final FileNotFoundException e) {
+                throwPathOrBranchNotFound(resolvedDirectory, branchRef);
+                return null;
+            }
+        });
+    }
+
+    /**
+     * Gets the current SHA for the given path from the given branch.
+     *
+     * @param path the path to the content
+     * @param branch the branch
+     * @return current sha for the given file, or empty optional
+     *
+     * @throws IOException if an I/O error happens calling GitHub
+     */
+    public Optional<String> getContentSha(final String path, final String branch) throws IOException, FlowRegistryException {
+        final String resolvedPath = getResolvedPath(path);
+        final String branchRef = BRANCH_REF_PATTERN.formatted(branch);
+        LOGGER.debug("Getting content SHA for [{}] from branch [{}] in repo [{}] ", resolvedPath, branch, repository.getName());
+
+        return execute(() -> {
+            try {
+                final GHContent ghContent = repository.getFileContent(resolvedPath, branchRef);
+                return Optional.of(ghContent.getSha());
+            } catch (final FileNotFoundException e) {
+                LOGGER.warn("Unable to get content SHA for [{}] from branch [{}] because content does not exist", resolvedPath, branch);
+                return Optional.empty();
+            }
+        });
+    }
+
+    /**
+     * Deletes the contents for the given file on the given branch.
+     *
+     * @param filePath the file path to delete
+     * @param commitMessage the commit message for the delete commit
+     * @param branch the branch to delete from
+     * @return the deleted content
+     *
+     * @throws IOException if an I/O error happens calling GitHub
+     * @throws FlowRegistryException if a non I/O error happens calling GitHub
+     */
+    public GHContent deleteContent(final String filePath, final String commitMessage, final String branch) throws FlowRegistryException, IOException {
+        final String resolvedPath = getResolvedPath(filePath);
+        LOGGER.debug("Deleting file [{}] in repo [{}] on branch [{}]", resolvedPath, repository.getName(), branch);
+        return execute(() -> {
+            try {
+                GHContent ghContent = repository.getFileContent(resolvedPath);
+                ghContent.delete(commitMessage, branch);
+                return ghContent;
+            } catch (final FileNotFoundException fnf) {
+                throwPathOrBranchNotFound(resolvedPath, branch);
+                return null;
+            }
+        });
+    }
+
+    private String getResolvedPath(final String path) {
+        return repoPath == null ? path : repoPath + "/" + path;
+    }
+
+    private void throwPathOrBranchNotFound(final String path, final String branch) throws FlowRegistryException {
+        throw new FlowRegistryException("Path [" + path + "] or Branch [" + branch + "] not found");
```

Comment:
Depending on which part of the path is not existing (bucket, flow, version), there are specific `FlowRegistryException` subclasses for the case. Also as the new concept "Branch" was added, it might worth to add a specific exception for that as well.

Reply:
Thanks for pointing out the sub-classes, I wasn't aware of those. There might be a few cases where we could leverage one of them, but I don't think we can use it here because we don't actually know which part doesn't exist. It could be any one of the bucket directory, the snapshot file name, or the branch. Unless we make a lot of individual requests to check each piece, but I don't want to do that because then we end up making lots of remote calls to GitHub for one single operation, and during background synchronize of many versioned PGs, it will end up being thousands of calls to GitHub.